### DISCUSSION TITLE
Timelimit and MipGap using CyLP interface for CBC

### DISCUSSION THREAD

**Discussion by marcobonelli** [coin-or/CyLP#127] [Q&A] [answered]
I have a model that, in a short time period, finds quality solutions but takes a long time to find and prove its optimality.

To work around this situation, I'm trying to activate the MipGap and/or the TimeLimit parameters, but I can't find documentation or examples using them in CyLP.

Does anyone know how to activate them? I know how to do this on CPLEX and Gurobi, but I'm not finding how to do it on CyLP for CBC.

**Comment by tkralphs** [comment, marked as answer]
This is a good question. I poked around a bit and I could not see a way to pass in arbitrary parameters (I'm not the original author of CyLP). I'm kind of surprised that these particular parameters don't seem to be supported, though. It shouldn't be that hard to hack something up if you're comfortable modifying the source code of CyLP. This is where the parameters would get passed:
https://github.com/coin-or/CyLP/blob/1769ca681c7a3f004ad24f0ceae15f9a76fc3e95/cylp/cpp/ICbcModel.cpp#L34-L37
Right now, they're hard-coded, but it would not be hard to make the `solve()` call of CyCbcModel just take a command-line string that would set whatever parameters you like and pass them in here. 
https://github.com/coin-or/CyLP/blob/b92ba0af3e853ccc0c47eaafe72abdd46837f827/cylp/cy/CyCbcModel.pyx#L145-L150
Feel free to open an issue if you're not comfortable doing this yourself, but I don't know when I'll have the bandwidth to look at it. If you want some pointers how to do it, I can give some more details. I would be happy to have a PR.

**Reply by marcobonelli** [reply]
Thanks for your answer.

As I don't think I have enough knowledge to do this without direction, I would be very grateful if you can give me more details on how I could do this modification in the source code.

As I am using CyLP for CBC in a specific project, I think trying to make this modification might be the quickest way to achieve this.

**Reply by tkralphs** [reply]
Here is what I think the steps should be roughly, but since I'm just saying this off the top of my head, it's possible this list might need to be tweaked.

- Clone this repo.
- Modify `CyCbcModel.pyx` to add a dictionary as an argument to the `solve()` method.
- Modify `ICbcModel.cpp::cbcMain()` to take an argument `const char** argv`.
- Modify the call to `cbcMain` in `solve()`, so that the dictionary gets converted and passed in to `cbcMain()`
- Comment the line that currently hard-codes the contents of `argv`

Then you just need to build CyLP except that you will need to run Cython to incorporate these changes.

- Install Cython (`pip install cython`)
- Set an environment variable `CYLP_USE_CYTHON` (no value needed, in `bash`, do `export CYLP_USE_CYTHON=`)
- Finally, do `pip install .` in the directory containing `setup.py`

Hopefully, this all goes smoothly, but if you want to get a bit more information about what's happening step-by-step, you can replace the `pip install .` above with separate build and install. This will also give you much more verbose output. 

- `python setup.py build`
- `python setup.py install`

In the first step, you should first see a bunch of output indicating that Cython is re-generating the auto-generated files. After that, it will do a build as normal.
